hazelcast/src/main/java/com/hazelcast/query/impl/CachedQueryEntry.java

```diff
@@ -102,7 +102,10 @@ protected Object getTargetObject(boolean key) {
             }
         } else {
             if (valueObject == null) {
-                if (valueData.isPortable()) {
+                // valueObject and valueData are null when the default constructor is used and init() is not yet called
+                if (valueData == null) {
+                    return null;
+                } else if (valueData.isPortable()) {
```

Comment:
Seems returning null is hiding the underlying issue. 

Passing `null` as an `oldValue` to the event after `map#set` can cause this NPE. Also it seems expected that one of valueData or valueObject should be set to a non-null value (at least for `getTargetObject` method).